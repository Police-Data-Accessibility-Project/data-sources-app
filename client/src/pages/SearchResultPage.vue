<template>
	<main data-test="search-results-page" class="flex flex-col p-8 h-auto">
		<div>
			<h1>Data Sources Search results</h1>
			<p data-test="search-results-section-header-p" class="text-2xl">
				Searching for <span class="font-semibold">"{{ searchTerm }}"</span> in
				<span class="font-semibold">"{{ location }}"</span>.
				<span v-if="searched && count > 0" data-test="search-results-count"
					>Found {{ getResultsCopy() }}.</span
				>
			</p>

			<Button class="my-4" intent="secondary" @click="() => $router.push('/')">
				<i class="fa fa-plus" /> New search
			</Button>
		</div>

		<p v-if="!searched" component="p" data-test="loading" :span-column="3">
			Loading results...
		</p>

		<p v-else-if="searched && count === 0" data-test="no-search-results">
			No results found.
		</p>

		<div v-else>
			<p class="text-xl max-w-full">
				If you don't see what you need,
				<a
					href="https://airtable.com/shrbFfWk6fjzGnNsk"
					data-test="search-results-request-link"
				>
					make a request&nbsp;<i class="fa fa-external-link" />
				</a>
			</p>
			<p class="text-xl max-w-full">
				To see these results in a table,
				<a href="https://airtable.com/shrUAtA8qYasEaepI">
					view the full database&nbsp;<i class="fa fa-external-link" />
				</a>
			</p>
		</div>

		<div data-test="search-results">
			<section
				v-for="section in uiShape"
				:key="section.header"
				data-test="search"
				class="mt-8 p-0 w-full"
			>
				<h2 class="section-subheading w-full">
					{{ section.header }}
				</h2>

				<div class="grid pdap-grid-container-column-3 gap-4">
					<ErrorBoundary
						v-for="result in [...getAllRecordsFromSection(section)]"
						:key="result.type"
					>
						<SearchResultCard
							data-test="search-results-cards"
							:data-source="result"
						/>
					</ErrorBoundary>
				</div>
			</section>
		</div>
	</main>
</template>

<script>
import { Button } from 'pdap-design-system';
import SearchResultCard from '../components/SearchResultCard.vue';
import ErrorBoundary from '../components/ErrorBoundary.vue';
import axios from 'axios';
import pluralize from '../util/pluralize';
import { SEARCH_RESULTS_UI_SHAPE } from '../util/pageData';

export default {
	name: 'SearchResultPage',
	components: {
		Button,
		ErrorBoundary,
		SearchResultCard,
	},
	data: () => ({
		count: 0,
		searched: false,
		searchResult: {},
		searchTerm: '',
		location: '',
		uiShape: {},
	}),
	mounted: function () {
		this.searchTerm = this.$route.params.searchTerm;
		this.location = this.$route.params.location;
		this.search();
	},
	methods: {
		getAllRecordsFromSection(section) {
			return section.records.reduce((acc, cur) => {
				return [...acc, ...this.searchResult[cur.type]];
			}, []);
		},
		getResultsCopy() {
			return `${this.count} ${pluralize('result', this.count)}`;
		},
		async search() {
			const url = `${
				import.meta.env.VITE_VUE_API_BASE_URL
			}/search-tokens?endpoint=quick-search&arg1=${this.searchTerm}&arg2=${
				this.location
			}`;

			try {
				const res = await axios.get(url);

				// Format results into object keyed by record_type
				const resultFormatted = res.data.data.reduce((acc, cur) => {
					return {
						...acc,
						[cur.record_type]: Array.isArray(acc[cur.record_type])
							? [...acc[cur.record_type], cur]
							: [cur],
					};
				}, {});

				// Modify ui shape object to exclude any sections / data sources that do not have records returned by API
				this.uiShape = SEARCH_RESULTS_UI_SHAPE.reduce((acc, cur) => {
					const recordsFiltered = cur.records.filter(
						(record) => resultFormatted[record.type],
					);
					return recordsFiltered.length > 0
						? [...acc, { header: cur.header, records: recordsFiltered }]
						: acc;
				}, []);

				// Set data and away we go
				this.searchResult = resultFormatted;
				this.count = res.data.count;
			} catch (error) {
				console.error(error);
			} finally {
				this.searched = true;
			}
		},
	},
};
</script>
