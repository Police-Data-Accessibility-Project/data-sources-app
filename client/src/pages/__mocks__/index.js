export const resultsMock = {
	count: 3,
	data: [
		{
			agency_name: 'Cicero Police Department - IN',
			agency_supplied: false,
			airtable_uid: 'rec7edsmcuVQwMMXG',
			coverage_end: null,
			coverage_start: '2017-01-01',
			data_source_name: 'Calls for Service for Cicero Police Department - IN',
			description: null,
			municipality: 'Bridgeview',
			name: 'Cicero Police Department - IN',
			record_format: '[]',
			record_type: 'Calls for Service',
			source_url:
				'https://cityprotect.com/agency/cc0d9a3d-50b2-4424-ae25-5699ba6eaff9',
			state_iso: 'IN',
		},
		{
			agency_name: 'Chicago Police Department - IL',
			agency_supplied: true,
			airtable_uid: 'recv9fMNEQTbVarj2',
			coverage_end: null,
			coverage_start: '2019-01-01',
			data_source_name: 'Calls for Service for Chicago Police Department - IL',
			description: null,
			municipality: 'Chicago',
			name: 'Chicago Police Department - IL',
			record_format: '[]',
			record_type: 'Calls for Service',
			source_url:
				'https://informationportal.igchicago.org/911-calls-for-cpd-service/',
			state_iso: 'IL',
		},
		{
			agency_name: 'Chicago Police Department - IL',
			agency_supplied: false,
			airtable_uid: 'recv9fMNEQTbVarj2',
			coverage_end: null,
			coverage_start: '2018-12-18',
			data_source_name: '311 Calls for City of Chicago',
			description:
				'311 Service Requests received by the City of Chicago. This dataset includes requests created after the launch of the new 311 system on 12/18/2018 and some records from the previous system, indicated in the LEGACY\\_RECORD column.\n\nIncluded as a Data Source because in some cities 311 calls lead to police response; that does not appear to be the case in Chicago.\n',
			municipality: 'Chicago',
			name: 'Chicago Police Department - IL',
			record_format: "['CSV', 'XML', 'RDF', 'RSS']",
			record_type: 'Calls for Service',
			source_url:
				'https://data.cityofchicago.org/Service-Requests/311-Service-Requests/v6vf-nfxy',
			state_iso: 'IL',
		},
	],
};

export const dataSourceMock = {
	name: 'Calls for Service for Chicago Police Department - IL',
	submitted_name: 'Chicago Police Department',
	description: null,
	record_type: 'Calls for Service',
	source_url:
		'https://informationportal.igchicago.org/911-calls-for-cpd-service/',
	last_cached: '01/03/2024',
	agency_supplied: true,
	supplying_entity: null,
	agency_originated: null,
	originating_entity: null,
	agency_aggregation: null,
	coverage_start: '2019-01-01',
	coverage_end: null,
	source_last_updated: null,
	retention_schedule: null,
	detail_level: 'Summarized totals',
	number_of_records_available: null,
	size: null,
	access_type: ['Web page'],
	data_portal_type: null,
	record_format: null,
	update_frequency: 'Bi-weekly',
	update_method: null,
	tags: null,
	readme_url: null,
	scraper_url: null,
	data_source_created: '2022-09-28',
	airtable_source_last_modified: '2023-11-08',
	url_status: 'ok',
	rejection_note: null,
	last_approval_editor:
		'{"id": "usrtLIB4Vr3jTH8Ro", "email": "josh.chamberlain@pdap.io", "name": "Josh Chamberlain"}',
	agency_described_submitted: 'Chicago city, Illinois',
	agency_described_not_in_database: null,
	approval_status: 'approved',
	record_type_other: null,
	data_portal_type_other: null,
	records_not_online: null,
	data_source_request: null,
	url_button: null,
	tags_other: null,
	access_notes: null,
	homepage_url: 'https://home.chicagopolice.org/',
	count_data_sources: 21,
	agency_type: 'law enforcement/police',
	multi_agency: null,
	jurisdiction_type: 'local',
	state_iso: 'IL',
	municipality: 'Chicago',
	zip_code: '60616',
	county_fips: '17031',
	county_name: ['Cook'],
	lat: 41.85861,
	lng: -87.62802,
	data_sources: [
		'recs7gYAGAMWEhLYC',
		'recyR4IVgqyIekzYB',
		'recBMxu4UcHCqjsWQ',
		'recqXzWtmhfV9z8Av',
		'recc5QWeCMGL17Ab6',
		'recD9bNxneanhBBDH',
		'recPUGM4OEopoRAXB',
		'rec10Mts7CmMPzcnx',
		'recTj2zTdu7Jp8oX9',
		'rec00T2YLS2jU7Tbn',
		'recpOsfsTV4bKmVfa',
		'recipwTdMN0jRADrt',
		'recjFk7X0KofjTPGK',
		'rec0Lqwrj6blSbkuz',
		'recYyAY8XVQJJVde6',
		'recgwPuRTwjAMUZeO',
		'recOLt1yQGlgrO5jQ',
		'recdAte1fo5FBruE7',
		'recUGIoPQbJ6laBmr',
		'recTw75rItd827L0r',
		'rechI06qD4od759xT',
		'recodGxvuxiTXPnOL',
	],
	no_web_presence: null,
	airtable_agency_last_modified: '2023-05-16',
	data_sources_last_updated: '2023-04-06',
	approved: true,
	rejection_reason: null,
	agency_created: '2022-08-18',
	county_airtable_uid: 'recuhY2ud60V41j0w',
	defunct_year: null,
	data_source_id: 'rec00T2YLS2jU7Tbn',
	agency_id: 'recv9fMNEQTbVarj2',
	agency_name: 'Chicago Police Department - IL',
};
