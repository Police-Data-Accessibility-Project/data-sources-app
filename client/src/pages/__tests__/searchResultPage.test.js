import SearchResultPage from '../SearchResultPage.vue';
// import { FlexContainer } from "pdap-design-system";
import { flushPromises, mount } from '@vue/test-utils';
import { describe, expect, vi, test, beforeEach, beforeAll } from 'vitest';
import axios from 'axios';
import { resultsMock } from '../__mocks__';
import { nextTick } from 'vue';

vi.mock('axios');

const $routeMock = {
	params: {
		searchTerm: 'calls',
		location: 'Cook',
	},
};

let wrapper;

beforeAll(() => {
	import.meta.env.VITE_VUE_APP_BASE_URL = 'https://data-sources.pdap.io';
});

describe('SearchResultPage renders with data', () => {
	beforeEach(() => {
		wrapper = mount(SearchResultPage, {
			global: {
				mocks: {
					$route: $routeMock,
				},
				stubs: {
					FlexContainer: {
						template: '<main><slot /></main>',
					},
				},
			},
		});

		axios.get.mockResolvedValue({
			data: resultsMock,
		});
	});

	test('Calls API and renders search results', async () => {
		const searchTerm = wrapper.vm.searchTerm;
		const location = wrapper.vm.location;

		await wrapper.vm.search();
		await flushPromises();

		// TODO: figure out why this is called twice
		expect(axios.get).toHaveBeenCalledTimes(2);
		expect(axios.get).toHaveBeenCalledWith(
			`https://data-sources.pdap.io/search-tokens?endpoint=quick-search&arg1=${searchTerm}&arg2=${location}`,
		);

		await nextTick();

		expect(wrapper.find('[data-test="search-results-page"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('renders search results section header properly', () => {
		axios.get.mockResolvedValue({
			data: resultsMock,
		});

		const searchTerm = wrapper.vm.searchTerm;
		const location = wrapper.vm.location;
		const count = wrapper.vm.searchResult.count;

		expect(
			wrapper.get('[data-test="search-results-section-header-p"]').text(),
		).toBe(
			`Searching for "${searchTerm}" in "${location}". Found ${count} results.`,
		);
	});

	test('renders search result count properly', () => {
		// const count = wrapper.vm.searchResult.count;

		expect(wrapper.get('[data-test="search-results-count"]').text()).toBe(
			`Found ${wrapper.vm.getResultsCopy()}.`,
		);
	});

	test('request data link has correct href value', () => {
		expect(
			wrapper.get('[data-test="search-results-request-link"]').attributes()
				.href,
		).toBe('https://airtable.com/shrbFfWk6fjzGnNsk');
	});

	test('search results card count matches search results returned in data', () => {
		expect(wrapper.findAll('[data-test="search-results-cards"]').length).toBe(
			wrapper.vm.searchResult.count,
		);
	});
});

describe('SearchResultPage shows no results found when no data returns', () => {
	test('search results page shows no results properly', async () => {
		axios.get.mockResolvedValue({
			data: {
				count: 0,
				data: [],
			},
		});

		wrapper = mount(SearchResultPage, {
			global: {
				mocks: {
					$route: $routeMock,
				},
			},
		});

		await flushPromises();

		expect(wrapper.find('[data-test="no-search-results"]').exists()).toBe(true);
	});
});
