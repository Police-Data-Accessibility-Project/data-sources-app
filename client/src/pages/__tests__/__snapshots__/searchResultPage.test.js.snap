// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchResultPage renders with data > Calls API and renders search results 1`] = `
<main class="flex flex-col p-8 h-auto">
  <div>
    <h1>Data Sources Search results</h1>
    <p class="text-2xl"> Searching for <span class="font-semibold">"calls"</span> in <span class="font-semibold">"Cook"</span>. <span>Found 3 results.</span>
    </p>
    <button class="pdap-button pdap-button-secondary my-4">
      <i class="fa fa-plus"></i> New search
    </button>
  </div>
  <section class="pdap-grid-container pdap-grid-container-column-3 p-0 w-full md:max-w-[unset] lg:max-w-[unset]" style="grid-template-rows: auto auto 1fr;">
    <div class="pdap-grid-item pdap-grid-item-span-column-3">
      <p class="text-xl max-w-full"> If you don't see what you need, <a href="https://airtable.com/shrbFfWk6fjzGnNsk"> make a request <i class="fa fa-external-link"></i>
        </a>
      </p>
      <p class="text-xl max-w-full"> To see these results in a table, <a href="https://airtable.com/shrUAtA8qYasEaepI"> view the full database <i class="fa fa-external-link"></i>
        </a>
      </p>
    </div>
    <div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
      <h2 class="text-xl font-semibold line-clamp-2">Calls for Service for Cicero Police Department - IN</h2>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
      <div class="mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit">Calls for Service</div>
      <div class="search-result-agency">
        <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
        <p>Cicero Police Department - IN</p>
      </div>
      <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
      <p>2016–Unknown end </p>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
      <ul class="mb-4">
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">[</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">]</li>
      </ul>
      <div class="mt-auto flex flex-col">
        <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full"> Visit data source <i class="fa fa-external-link"></i>
        </button>
        <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
      </div>
    </div>
    <div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
      <h2 class="text-xl font-semibold line-clamp-2">Calls for Service for Chicago Police Department - IL</h2>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
      <div class="mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit">Calls for Service</div>
      <div class="search-result-agency">
        <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
        <p>Chicago Police Department - IL</p>
      </div>
      <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
      <p>2018–Unknown end </p>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
      <ul class="mb-4">
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">[</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">]</li>
      </ul>
      <div class="mt-auto flex flex-col">
        <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full"> Visit data source <i class="fa fa-external-link"></i>
        </button>
        <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
      </div>
    </div>
    <div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
      <h2 class="text-xl font-semibold line-clamp-2">311 Calls for City of Chicago</h2>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
      <div class="mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit">Calls for Service</div>
      <div class="search-result-agency">
        <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
        <p>Chicago Police Department - IL</p>
      </div>
      <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
      <p>12/17/2018–Unknown end </p>
      <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
      <ul class="mb-4">
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">[</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">C</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">S</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">V</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">,</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit"> </li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">X</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">M</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">L</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">,</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit"> </li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">R</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">D</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">F</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">,</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit"> </li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">R</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">S</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">S</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
        <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">]</li>
      </ul>
      <div class="mt-auto flex flex-col">
        <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full"> Visit data source <i class="fa fa-external-link"></i>
        </button>
        <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
      </div>
    </div>
  </section>
</main>
`;
